Add `lt` and `by` as keyword args to median (Resolves #19190)

ntdef opened the pull request:
Per request from #19190.

Commit:
Add `lt`, `by` and `rev` as keyword args to median

ararslan commented:
The `rev` keyword seems unnecessary because the median should be the same regardless of whether the data are sorted in ascending or descending order.

ntdef commented:
@ararslan _facepalms_ Very true.

Commit:
Remove `rev` keyword arg from median

The median is the same regardless of whether it is sorted ascending or
descending.

ntdef commented:
OK, I've updated the PR to remove the `rev` arg

ararslan commented on base/statistics.jl:
```diff
@@ -571,15 +571,20 @@ function median!{T}(v::AbstractVector{T})
+applied to each element before comparison. The `lt` keyword allows providing a
+custom "less than" function for sorting prior to taking the median.
+"""
+median{T}(v::AbstractArray{T}; lt=isless, by=identity, rev::Bool=false) = median!(copy!(Array{T,1}(_length(v)), v), lt=lt, by=by)
```
You'll have to remove `rev` from here as well

ntdef replied:
Updated. See commit below.

Commit:
Fix `rev` arg straggler

simonbyrne commented:
Thanks for this. Have you done any performance checks? There is a chance we might hit a performance regression due to the keyword arguments (note how `Base.Sort` uses `Order.ord` function).

The best approach would be to add something to [BaseBenchmarks.jl](https://github.com/JuliaCI/BaseBenchmarks.jl).

ntdef commented:
@simonbyrne Definitely worthwhile to check. I'll add a benchmark and see how it looks.

andreasnoack commented:
Are we sure about how a `by`-median is supposed to work? I'm wondering if the use of the arithmetic mean for even length samples makes sense when we use `by`. A median `m` should satisfy
```
P(X ≥ m) ≥ 1/2 and P(X ≤ m) ≥ 1/2.
```
I'd guess that a `by`-median should then satisfy
```
P(by(X) ≥ by(m)) ≥ 1/2 and `P(by(X) ≤ by(m)) ≥ 1/2.
```
Is that right? If so then `median([1,-2,3,-4], by=abs) === 0.5` is wrong. More generally, it might be possible to compute medians for which the mean is not well defined (as mentioned in A followup on https://github.com/JuliaLang/julia/issues/19190#issuecomment-257885325). Another example is a vector of tuples and `by=first`.

stevengj commented:
I would think that `median(a, by=f)` should be equivalent to (but faster than) `median(map(f, a))`. In which case we should have `median([1,-2,3,-4], by=abs) == 2.5`.

cstjean commented:
> I would think that `median(a, by=f)` should be equivalent to (but faster than) `median(map(f, a))`.

`median(f(x) for x in a)`? 

I like @andreasnoack's proposal. `maximum(students, by=s->s.height)` should be a `Student` object: the tallest student in the classroom. `median` should be similarly defined if it supports `by` at all.

ntdef commented:
I agree with @andreasnoack's definition of `by` for the median. Admittedly, `median([1,-2,3,-4], by=abs) === 0.5` is a poor example of how one might want to use `by` in practice. The problem, of course, is defining what we mean by `middle` in the case of an even number of arguments.

One approach is to only allow use of the 'by' keyword if you also define a 'middle' function. Borrowing @cstjean's Student objects, consider the following example. 
```
immutable Student
    name::String
    height::Int64
end

classroom = [
    Student("Alice",  153),
    Student("Bob",    172),
    Student("Claire", 160),
    Student("Daniel", 155),
]

Base.isless(a::Student, b::Student) = isless(a.height, b.height)
Base.middle(a::Student, b::Student) = isless(a, b) ? a : b

ans = median(classroom, by=x->x.height)
println(ans.name)
# Daniel
```
"Daniel" and "Claire" are both possible answers to `P(by(X) ≥ by(m)) ≥ 1/2 and P(by(X) ≤ by(m)) ≥ 1/2`. To resolve the ambiguity and get one answer, we have to define what `middle` means for this type. I could also imagine a possible answer where `middle` returns a 2-tuple with both options. The way this would work in practice is before computing a result for `median`, we check whether `middle` is defined for the argument type and raise an error if not.

stevengj commented:
@cstjean, one problem is that
```
julia> median(abs(x) for x in [1,-2,3,-4])
ERROR: MethodError: no method matching median(::Base.Generator{Array{Int64,1},Base.#abs})
```
We don't have a generic `median` function for iterables, presumably because the fast algorithms require random access.  Of course, we could define a `median` function for `Generator{T<:AbstractArray}` or use an "indexable" trait.

andreasnoack commented:
*If* we can come up with some solution for the `middle` problem then I think that `foo(median(x, by = foo))` should also be sufficiently efficient.

stevengj commented:
@andreasnoack, I'm just not sure that it's worth it.  I understand the application for `median(map(f, x))`.  I don't understand the application for the proposed `median(x, by=f)`.   Why not just define `median(x, map=f)` if we don't like using `by` for this purpose, and/or define it for generators on arrays?

ararslan commented:
@stevengj's suggestion of `median(x, map=f)` seems reasonable, though for consistency with `sum`, `mean`, etc. it would probably be better as `median(f, x)`.

andreasnoack commented:
I'd be fine with `median(f,x)` but I kind of like that `median` would match `sort` here and I also could imagine it being useful for records similar to @cstjean's example. @shashi just proposed the obvious that the `middle` function is just a user-supplied function, so maybe something like
```jl
median(x, by = identity, tie = defaulttie(eltype(x)))
defaulttie{T<:Number}(x::Type{T}) = middle(x,y)
defaulttie(::) = (x,y) -> x ## doesn't matter which we choose
```

ararslan commented:
Since `tie` already doesn't match the `sort` API, you could mix approaches with `median(f, x, tie=middle)` (or `defaulttie` as you've defined or whatever).

andreasnoack commented:
`median(f, x, tie=middle)` is less general than `median(x, by=f)` because in the latter case you are not forced to apply `f` on the result, i.e. get the student as the result instead of his height.

StefanKarpinski commented:
How about just defining `median(v, by=f)` in the case of an even length `v` to take the middle element which sorts first? If you want the other one, you can just do `median(v, by=f, rev=true)`?

ararslan commented:
@StefanKarpinski That could also be accomplished with a `tie` argument, i.e. using `tie=first` or `tie=last`. Having a `rev` argument in `median` seems unintuitive to me, but maybe that's just me.

StefanKarpinski commented:
The point is that it matches the existing API for sorting-related functions instead of introducing a new median-specific API that isn't applicable to anything else. It's also really simple.

ararslan commented:
That's fair

ararslan commented:
Using the behavior you suggest, would it still be possible to get the current behavior using `middle`? If so, how?

StefanKarpinski commented:
No, I guess that's the drawback – you can only get one median or the other, not some function of the two. The current behavior would apply if and only if `by == identity`.

ararslan commented:
So it would use `middle` if `by == identity`?

StefanKarpinski commented:
I don't want to introduce this new `middle` function at all.

ararslan commented:
?? It's already there...

StefanKarpinski commented:
Ah, I didn't realize that we already had a `middle` function.

ntdef commented:
`middle` is already exported in the standard lib. It's documented and everything.

ararslan commented:
Yeah. Currently in the case of even length, we take `middle(x, y)`, which is just `(x + y) / 2`. I think doing away with that behavior in `median` now may have major ramifications in the stats ecosystem.

ntdef commented:
I like the `rev` approach, except that it clashes with how `median` currently works by default on a list of numbers (by taking the average) and therefore might be confusing if that's what you're used to. 

The one thing to keep in mind is that whatever API we choose for `median` should probably be applicable to `quantile`.

nalimilan commented:
I'd rather call `middle` by default, but provide a way not to call it if needed. I also don't really see the point of using `rev` for consistency with `sort`, since the consequences of this argument are really different here. There's no specific reason to choose the first element rather than the last one. Better have `tie` to be more explicit.

A possible API would be `tie=middle`, with `first` and `last` as common possibilities. The passed function would just be called with the two elements in the middle of the vector (for even lengths).

StefanKarpinski commented:
Shouldn't `middle` be called with one argument even when the length is odd?

ararslan commented:
It is currently

StefanKarpinski commented:
If you call `first` on the middle two values when there's a tie and `middle` on the middle value when there's no tie, that's going to lead to type instability, e.g. when operating on an `Int` array.

ararslan commented:
Isn't that kind of an opt-in type instability though, since you're knowingly passing a function that will yield a return type other than what `middle` would? (Or maybe that's nonsense...)

nalimilan commented:
Stefan is right that the function would need to be called with one argument even when there's no tie. We don't want the type to change depending on the length of the input. Fortunately, `middle(x)` already works.

StefanKarpinski commented:
I guess, although then you'll probably want to do `median(v, by=f, tie=(x,y)->middle(x))` instead of `tie=first` most of the time. Also, `first` doesn't really work since it takes a collection and has no two-argument form. One option would be to define `first(x,y)` to return the same type as `middle(x,y)` would but then `first(x)` won't return the right type unless we change it.

ararslan commented:
Internally we can pass the pair in the case of even length as an array to the given `tie` function, which would work for `first`, `last`, and `middle` (among others). Then we can call `middle` on the result before it's returned.

andreasnoack commented:
Why not just call a two argument function in both the even and odd case, i.e. something like `foo(x[(n - 1) >> 1 + 1], x[n >> 1 + 1])` for all `n`. Seems like a much simpler way of achieving type stability.

StefanKarpinski commented:
I like @andreasnoack's suggestion. Calling the function `tie` is a little misleading then though.

nalimilan commented:
Not sure what a more appropriate name could be. `choose`? `result`?

StefanKarpinski commented:
Maybe `middle` so the default is `middle=middle`. It's a bit weird to write `middle=first` though.

andreasnoack commented:
@ntdef I think we have almost reached a consensus here so would you be able to update the PR? I'd be okay with `middle=middle` since I cannot come up with something better. I think the default for `Number` elements should be `middle=middle` and otherwise `middle=first` and the function should get called in both the even and odd case.

ntdef commented:
Certainly! I'll patch it up this afternoon.
On Wed, Nov 30, 2016 at 8:07 AM Andreas Noack <notifications@github.com>
wrote:

> @ntdef <https://github.com/ntdef> I think we have almost reached a
> consensus here so would you be able to update the PR? I'd be okay with
> middle=middle since I cannot come up with something better. I think the
> default for Number elements should be middle=middle and otherwise
> middle=first and the function should get called in both the even and odd
> case.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/19359#issuecomment-263881567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJEEccdHenGVhHXpOJjFejmWLxZedIudks5rDYMkgaJpZM4K2B9Z>
> .
>

stevengj commented:
Can you also add `median(f, x)` (for `median(map(f, x))`)?

ntdef committed:
WIP Add `middle` as an argument to `median`

This is a work-in-progress commit, showing an issue with `median` throwing
strange UnDefVar errors when running `make test-statistics`.

ntdef commented:
I added `middle` as an argument but I seem to be getting these `UnDefVar` errors that disappear when I test the function out in the repl. Could I maybe get a fresh set of eyes on the last commit?

andreasnoack commented:
In this process, I seem to have forgotten my original comment https://github.com/JuliaLang/julia/pull/19359#issuecomment-261580952. It seems to be impossible to combine a `by` keyword with using the average of the two middle observations in the even case. For `by=abs` the result is nonsense. It would work if just used `first` or `last` but the average solution is taught in any first stats course in the world so it is probably not a good idea to deviate here. I'd suggest that we focus on just the `median(f, x)` solution for now because it seems to be unambiguous and there is a real demand for the functionality.

StefanKarpinski commented on test/statistics.jl:
```diff
@@ -32,6 +32,23 @@ ismore(x,y) = !isless(x,y)
+@test ans.name == "Daniel"
+# ans = median(classroom, by=x->x.height)
+# @test ans.name == "Daniel"
+
```
Maybe test that if you add the `rev=true` option you get Claire?

StefanKarpinski commented on base/statistics.jl:
```diff
@@ -573,18 +573,23 @@ function median!{T}(v::AbstractVector{T}; lt=isless, by=identity)
+median{T}(v::AbstractArray{T}; lt=isless, by=identity, middle=middle) = median!(copy!(Array{T,1}(_length(v)), v), lt=lt, by=by, middle=middle)
+
+# median(v::abstractarray{number}; lt=isless, by=identity, middle=middle) =
+#     median!(copy!(Array{Number,1}(_length(v)), v), lt=lt, by=by, middle=middle)
```
What's the commented out method?

ntdef commented:
I agree; the `by` argument is not shaking out to be coherent enough or useful enough to warrant adding. 

The one use-case that I think it makes sense is when you want to find the *object* that is the median, as in my classroom example. A case like that could be covered by a function like `indmedian`, similar to `indmax` or `indmin` that doesn't suffer from the same ambiguity, and for which it would be easier to define a tie-braking function. We already have `indmax` and `indmin` so `indmedian` should make sense intuitively..

oscardssmith commented:
Superseded by https://github.com/JuliaLang/julia/pull/30329
